### PR TITLE
Squeal Version 0.6

### PR DESCRIPTION
**Pull request by echatav**
New version of Squeal with many features; see `RELEASE NOTES.md`. Would be lovely if folks would take a look at this. Looking to nitpicks in documentation or any mistakes small or large that you can find.

### CODE REVIEW

File: RELEASE NOTES.md [hunk 1]
```diff
@@ -1,5 +1,288 @@
...
+PG Foo :: PGType
+= 'PGjsonb
+
+>>> printSQL (inline (Baz 'a' "aaa"))
+('{"tag":"Baz","contents":["a","aaa"]}' :: jsonb)
+```
+
+One thing to notice about `ToParam db` is that it has an
+extra parameter `db` that the other classes don't have. That's
+because for some types, such as arrays and composites; you
```
Comment (simfleischman) [inline RELEASE NOTES.md:200]:
```suggestion
because for some types, such as arrays and composites, you
```

File: RELEASE NOTES.md [hunk 2]
```diff
@@ -1,5 +1,288 @@
...
+
+Lots of other things changed. `Literal` and `literal` are now called
+`Inline` and `inline`. `ColumnConstraint` is called `Optionality`.
+Squeal 0.6 adds support for domain types. It more carefully types
+`CREATE _ IF NOT EXISTS` and `DROP _ IF EXISTS` definitions. The
+`Exception` type was refactored to remove `Maybe`s and new pattern
+synonyms were defined to easily match on a few common SQL errors.
+`VarChar` and `FixChar` types were added with smart constructors.
+Many bugs were fixed. Also, many more tests were added and
+a new benchmark suite. A lot more things were changes that I've
```
Comment (simfleischman) [inline RELEASE NOTES.md:279]:
```suggestion
a new benchmark suite. A lot more things were changed that I've
```

**Comment by ysangkok** [comment]
Some preprocessor error while building on GHC 8.10:

Running 1 test suites...
Test suite doctest: RUNNING...

```
src/Squeal/PostgreSQL/Type/PG.hs:278:0: error:
     error: missing binary operator before token "("
     #if MIN_VERSION_generics_sop(0,5,0)
     
    |
278 | #if MIN_VERSION_generics_sop(0,5,0)
    | ^

src/Squeal/PostgreSQL/Type/PG.hs:1:1: error:
    `cc' failed in phase `C pre-processor'. (Exit code: 1)
  |
1 | {-|
  | ^
Test suite doctest: FAIL
```

**Comment by echatav** [comment]
Oh no! Were there some changes to `CPP` GHC-8.10? Looks like stackage doesn't have a nightly with it yet.

**Comment by tuomohopia** [comment]
I didn't spot any changes to `CPP` in the GHC 8.10.1  release notes. 

@echatav Have you seen this?
https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/native-cpp

From what I understand, the compiler currently uses the system's own C-preprocessor. Maybe his computer's version is different than yours / than what it should be?

**Comment by ysangkok** [comment]
I deleted my GHC 8.10 environment and I don't have space for it right now, but there are issues/pull requests for base 4.14 compat for transitive dependencies, so I'll wait for those before testing again:

* haskell-hvr/uuid#47
* haskell-hvr/uuid#50
* haskell-hvr/uuid#51
* serokell/haskell-with-utf8#4

**Comment by echatav** [comment]
@ysangkok Did you happen to check that it was working for you with earlier versions of GHC?

**Comment by ysangkok** [comment]
> @ysangkok Did you happen to check that it was working for you with earlier versions of GHC?

I have just tested with GHC 8.8 right now. It does build, but the test suite does not pass:

    Exceptions
      should be thrown for constraint violation FAILED [1]
      should be rethrown for constraint violation in a transaction FAILED [2]
    Pools
      should manage concurrent transactions FAILED [3]
    Ranges
      should correctly decode ranges FAILED [4]
    Parameters
      should run queries that don't reference all their parameters FAILED [5]
    Composite types
      should be embeddible FAILED [6]

**Comment by echatav** [comment]
I'm going to merge this. I will probably release a version 0.6.1 to up the version of some dependencies (like `time` and `generics-sop`) later which will in any case obviate the need for CPP and as a byproduct will fix the GHC-8.10 issue.

**Pull request merged**